sys/senml/senml.c

```diff
@@ -0,0 +1,416 @@
+            }
+
+            ins_pos += chars_written;
+            buf_len -= chars_written;
+        }
+
+        if (curr_record->unit) {
+            chars_written = snprintf(&output[ins_pos], buf_len, "\"u\":\"%s\",",
+                                     curr_record->unit);
+
+            if (chars_written >= (ssize_t)buf_len || chars_written < 0) {
+                return -1;
+            }
+
+            ins_pos += chars_written;
+            buf_len -= chars_written;
+        }
+
+        if (curr_record->time != 0) {
+            chars_written = snprintf(&output[ins_pos], buf_len, "\"t\":%f,",
+                                     curr_record->time);
+
+            if (chars_written >= (ssize_t)buf_len || chars_written < 0) {
+                return -1;
+            }
+
+            ins_pos += chars_written;
+            buf_len -= chars_written;
+        }
+
+        if (curr_record->update_time != 0) {
+            chars_written = snprintf(&output[ins_pos], buf_len, "\"ut\":%f,",
+                                     curr_record->update_time);
+
+            if (chars_written >= (ssize_t)buf_len || chars_written < 0) {
+                return -1;
+            }
+
+            ins_pos += chars_written;
+            buf_len -= chars_written;
+        }
+
+        if (curr_record->value_sum != 0) {
+            chars_written = snprintf(&output[ins_pos], buf_len, "\"s\":%f,",
+                                     curr_record->value_sum);
+
+            if (chars_written >= (ssize_t)buf_len || chars_written < 0) {
+                return -1;
+            }
+
+            ins_pos += chars_written;
+            buf_len -= chars_written;
+        }
+
+        if (curr_record->value_type != SENML_TYPE_UNDEF) {
+            if (curr_record->value_type == SENML_TYPE_FLOAT) {
+                chars_written = snprintf(&output[ins_pos], buf_len, "\"v\":%f,",
+                                         curr_record->value.f);
+            }
+            else if (curr_record->value_type == SENML_TYPE_STRING) {
+                chars_written = snprintf(&output[ins_pos], buf_len, "\"vs\":\"%s\",",
+                                         curr_record->value.s);
+            }
+            else if (curr_record->value_type == SENML_TYPE_BOOL) {
+                chars_written = snprintf(&output[ins_pos], buf_len, "\"vb\":%s,",
+                                         curr_record->value.b ? "true" : "false");
+            }
+            else if (curr_record->value_type == SENML_TYPE_BINARY) {
+                chars_written = snprintf(&output[ins_pos], buf_len, "\"vd\":\"%s\",",
+                                         curr_record->value.d);
+            }
+
+            if (chars_written >= (ssize_t)buf_len || chars_written < 0) {
+                return -1;
+            }
+
+            ins_pos += chars_written;
+            buf_len -= chars_written;
+        }
+
+        if (buf_len == buf_len_old) {   // record was empty, no values inserted: remove '{'
```

Comment:
`senml_decode_json_s` sets `pack->num` to actual number of valid records. So `pack->num` may differ from the number of allocated records. Therefore, requiring caller of `senml_encode_json_s` to set `pack->num` to actual number of valid records may be acceptable.

Ignoring empty records my be less useful than expected. Typical application sends data repeatedly to a server and reuses a statically allocated pack. Caller must clear records before reusing the pack to make `senml_encode_json_s` ignore empty records correctly. Setting `pack->num` before calling `senml_encode_json_s` is easier than clearing the pack before reusing it. 

Example:
Suppose a RIOT node with a sensor device interrupting at irregular intervals. The node sends data to a server on timer, or when no space available in the buffer.
```C
#define N 10

static senml_record_t records[N]; 
static senml_pack_t pack;
static int next_record = 0;

static void init_pack() {
  pack.records = records;
  ... // init other fields and records.
}

void on_new_sensor_data_available(double time, double data) {
  if (next_record == N) {
    send_seml_to_server();
  }
  pack.records[next_record].time = time;
  pack.records[next_record].value.f = data;
  next_record++;
}

void on_timer() {
  send_seml_to_server();
}

void send_seml_to_server() {
  if (next_record == 0) {
    return;
  }
  pack.num = next_record;

  char buffer[1024];

  senml_encode_json_s(&pack, buffer, 1024);
  send_string_to_server(buffer);
  next_record = 0;
  // does not need to clear records here
}
```